# Data Collection
การนำเข้าข้อมูล และคำอธิบายจากไฟล์ README.txt

In [248]:
import pandas as pd

In [249]:
link = pd.read_csv("ml-latest-small/links.csv")
movie = pd.read_csv("ml-latest-small/movies.csv")
rating = pd.read_csv("ml-latest-small/ratings.csv")
tag = pd.read_csv("ml-latest-small/tags.csv")

## โครงสร้างไฟล์ข้อมูลลิงก์ (links.csv)
### ต้องไปดูในลิ้งอีกที
--------------------------------------


ตัวระบุที่สามารถใช้ลิงก์ไปยังแหล่งข้อมูลภาพยนตร์อื่นๆ มีอยู่ในไฟล์ links.csv แต่ละบรรทัดของไฟล์นี้หลังแถวส่วนหัวจะแสดงภาพยนตร์หนึ่งเรื่อง และมีรูปแบบดังต่อไปนี้:

movieId, imdbId, tmdbId

movieId คือตัวระบุสำหรับภาพยนตร์ที่ https://movielens.org ใช้ เช่น ภาพยนตร์เรื่อง Toy Story มีลิงก์ https://movielens.org/movies/1

imdbId คือตัวระบุสำหรับภาพยนตร์ที่ http://www.imdb.com ใช้ เช่น ภาพยนตร์เรื่อง Toy Story มีลิงก์ http://www.imdb.com/title/tt0114709/

tmdbId คือตัวระบุสำหรับภาพยนตร์ที่ https://www.themoviedb.org ใช้ เช่น ภาพยนตร์เรื่อง Toy Story มีลิงก์ https://www.themoviedb.org/movie/862

In [250]:
link.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

In [251]:
link["movieId"] = link["movieId"].astype('category')
link["imdbId"] = link["imdbId"].astype('category')
link["tmdbId"] = link["tmdbId"].astype('category')

In [253]:
link.describe(include='all')

movieId   imdbId  tmdbId
count      9742     9742  9734.0
unique     9742     9742  9733.0
top      193609  8391976  4912.0
freq          1        1     2.0

In [254]:
link.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   movieId  9742 non-null   category
 1   imdbId   9742 non-null   category
 2   tmdbId   9734 non-null   category
dtypes: category(3)
memory usage: 1.0 MB


## โครงสร้างไฟล์ข้อมูลภาพยนตร์ (movies.csv)
### ใช้อิงชื่อเรื่อง(title)และประเภท(genres)หนังกับไฟล์ที่เหลือด้วย movieId 
--------------------------------------

ข้อมูลภาพยนตร์จะอยู่ในไฟล์ `movies.csv` แต่ละบรรทัดของไฟล์นี้หลังแถวส่วนหัวแสดงถึงภาพยนตร์หนึ่งเรื่อง และมีรูปแบบดังต่อไปนี้:

movieId,title,genres

ชื่อภาพยนตร์สามารถป้อนด้วยตนเองหรือนำเข้าจาก <https://www.themoviedb.org/> และใส่ปีที่ออกฉายไว้ในวงเล็บ อาจมีข้อผิดพลาดและความไม่สอดคล้องกันในชื่อเหล่านี้

ประเภทต่างๆ เป็นรายการที่แยกจากกัน และเลือกจากรายการต่อไปนี้:

* แอ็กชัน
* ผจญภัย
* แอนิเมชัน
* สำหรับเด็ก
* ตลก
* อาชญากรรม
* สารคดี
* ดราม่า
* แฟนตาซี
* ฟิล์มนัวร์
* สยองขวัญ
* ดนตรี
* ลึกลับ
* โรแมนติก
* ไซไฟ
* ระทึกขวัญ
* สงคราม
* ตะวันตก
* (ไม่ระบุประเภท)

In [255]:
movie.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

In [256]:
#ทำให้หา unique ได้
movie["movieId"] = movie["movieId"].astype('category')

In [257]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype   
---  ------   --------------  -----   
 0   movieId  9742 non-null   category
 1   title    9742 non-null   object  
 2   genres   9742 non-null   object  
dtypes: category(1), object(2)
memory usage: 505.5+ KB


In [258]:
movie.describe(include='all')

movieId            title genres
count      9742             9742   9742
unique     9742             9737    951
top      193609  Saturn 3 (1980)  Drama
freq          1                2   1053

## โครงสร้างไฟล์ข้อมูลการให้คะแนน (ratings.csv)
### ให้ดาว 0.5-5 (เพิ่มทีละ 0.5)
-----------------------------------------

การให้คะแนนทั้งหมดอยู่ในไฟล์ ratings.csv แต่ละบรรทัดของไฟล์นี้หลังแถวส่วนหัวแสดงถึงการให้คะแนนภาพยนตร์หนึ่งเรื่องโดยผู้ใช้หนึ่งราย และมีรูปแบบดังนี้:

userId,movieId,rating,timestamp

บรรทัดภายในไฟล์นี้จะเรียงลำดับตาม userId ก่อน จากนั้นจึงเรียงลำดับตาม movieId ภายในผู้ใช้

การให้คะแนนจะคำนวณโดยใช้มาตราส่วน 5 ดาว โดยเพิ่มทีละครึ่งดาว (0.5 ดาว - 5.0 ดาว)

Timestamp แสดงถึงวินาทีนับตั้งแต่เที่ยงคืนตามเวลาสากลเชิงพิกัด (UTC) ของวันที่ 1 มกราคม ค.ศ. 1970

In [259]:
rating.head()

userId  movieId  rating  timestamp
0       1        1     4.0  964982703
1       1        3     4.0  964981247
2       1        6     4.0  964982224
3       1       47     5.0  964983815
4       1       50     5.0  964982931

In [260]:
#ทำให้หา unique ได้
rating["userId"] = rating["userId"].astype('category')
rating["movieId"] = rating["movieId"].astype('category')

In [261]:
rating.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype   
---  ------     --------------   -----   
 0   userId     100836 non-null  category
 1   movieId    100836 non-null  category
 2   rating     100836 non-null  float64 
 3   timestamp  100836 non-null  int64   
dtypes: category(2), float64(1), int64(1)
memory usage: 2.3 MB


In [262]:
rating.describe(include='all')

userId   movieId         rating     timestamp
count   100836.0  100836.0  100836.000000  1.008360e+05
unique     610.0    9724.0            NaN           NaN
top        414.0     356.0            NaN           NaN
freq      2698.0     329.0            NaN           NaN
mean         NaN       NaN       3.501557  1.205946e+09
std          NaN       NaN       1.042529  2.162610e+08
min          NaN       NaN       0.500000  8.281246e+08
25%          NaN       NaN       3.000000  1.019124e+09
50%          NaN       NaN       3.500000  1.186087e+09
75%          NaN       NaN       4.000000  1.435994e+09
max          NaN       NaN       5.000000  1.537799e+09

In [263]:
rating["userId"].unique()

[1, 2, 3, 4, 5, ..., 606, 607, 608, 609, 610]
Length: 610
Categories (610, int64): [1, 2, 3, 4, ..., 607, 608, 609, 610]

In [264]:
rating["movieId"].unique()

[1, 3, 6, 47, 50, ..., 160341, 160527, 160836, 163937, 163981]
Length: 9724
Categories (9724, int64): [1, 2, 3, 4, ..., 193583, 193585, 193587, 193609]

## โครงสร้างไฟล์ข้อมูลแท็ก (tags.csv)
### user เป็นคน tags หนังแต่ละเรื่องเองต่างกันไปในแต่ละคนและหนัง
-----------------------------------

แท็กทั้งหมดอยู่ในไฟล์ `tags.csv` แต่ละบรรทัดของไฟล์นี้หลังแถวส่วนหัวแสดงถึงหนึ่งแท็กที่ผู้ใช้หนึ่งรายนำไปใช้กับภาพยนตร์หนึ่งเรื่อง และมีรูปแบบดังต่อไปนี้:

userId,movieId,tag,timestamp

บรรทัดภายในไฟล์นี้จะเรียงลำดับตาม userId ก่อน จากนั้นจึงเรียงลำดับตาม movieId ภายในผู้ใช้

แท็กคือข้อมูลเมตาที่ผู้ใช้สร้างขึ้นเกี่ยวกับภาพยนตร์ โดยทั่วไปแล้วแต่ละแท็กจะเป็นคำหรือวลีสั้นๆ ความหมาย คุณค่า และวัตถุประสงค์ของแท็กนั้นๆ จะถูกกำหนดโดยผู้ใช้แต่ละราย

timestamp หมายถึงวินาทีนับตั้งแต่เที่ยงคืนตามเวลาสากลเชิงพิกัด (UTC) ของวันที่ 1 มกราคม ค.ศ. 1970

In [265]:
tag

userId  movieId               tag   timestamp
0          2    60756             funny  1445714994
1          2    60756   Highly quotable  1445714996
2          2    60756      will ferrell  1445714992
3          2    89774      Boxing story  1445715207
4          2    89774               MMA  1445715200
...      ...      ...               ...         ...
3678     606     7382         for katie  1171234019
3679     606     7936           austere  1173392334
3680     610     3265            gun fu  1493843984
3681     610     3265  heroic bloodshed  1493843978
3682     610   168248  Heroic Bloodshed  1493844270

[3683 rows x 4 columns]

In [266]:
tag["userId"] = tag["userId"].astype('category')
tag["movieId"] = tag["movieId"].astype('category')

In [267]:
tag.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3683 entries, 0 to 3682
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype   
---  ------     --------------  -----   
 0   userId     3683 non-null   category
 1   movieId    3683 non-null   category
 2   tag        3683 non-null   object  
 3   timestamp  3683 non-null   int64   
dtypes: category(2), int64(1), object(1)
memory usage: 115.5+ KB


In [268]:
tag.describe(include='all')

userId  movieId               tag     timestamp
count   3683.0   3683.0              3683  3.683000e+03
unique    58.0   1572.0              1589           NaN
top      474.0    296.0  In Netflix queue           NaN
freq    1507.0    181.0               131           NaN
mean       NaN      NaN               NaN  1.320032e+09
std        NaN      NaN               NaN  1.721025e+08
min        NaN      NaN               NaN  1.137179e+09
25%        NaN      NaN               NaN  1.137521e+09
50%        NaN      NaN               NaN  1.269833e+09
75%        NaN      NaN               NaN  1.498457e+09
max        NaN      NaN               NaN  1.537099e+09

In [269]:
tag["userId"].unique()

[2, 7, 18, 21, 49, ..., 573, 599, 600, 606, 610]
Length: 58
Categories (58, int64): [2, 7, 18, 21, ..., 599, 600, 606, 610]

# Data Cleaning

## เช็ค missing value

In [270]:
link.isnull().sum()

movieId    0
imdbId     0
tmdbId     8
dtype: int64

In [271]:
link.loc[link["tmdbId"].isnull()]

movieId  imdbId tmdbId
624      791  113610    NaN
843     1107  102336    NaN
2141    2851   81454    NaN
3027    4051   56600    NaN
5532   26587   92337    NaN
5854   32600  377059    NaN
6059   40697  105946    NaN
7382   79299  874957    NaN

### มี NaN ที่ link จริงแต่ปล่อยไว้ก่อน น่าจะไม่ใช้ทำโมเดล 

In [272]:
movie.isnull().sum()

movieId    0
title      0
genres     0
dtype: int64

In [273]:
rating.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

In [274]:
tag.isnull().sum()

userId       0
movieId      0
tag          0
timestamp    0
dtype: int64

## เช็ค Inconsistent

### จาก unique ที่ไม่เท่ากันระหว่าง movieId กับ title

In [275]:
movie.describe(include='all')

movieId            title genres
count      9742             9742   9742
unique     9742             9737    951
top      193609  Saturn 3 (1980)  Drama
freq          1                2   1053

In [276]:
#หา title ที่มี movieId ไม่เท่ากับ 1
not_oneId_title = movie["title"].value_counts()[movie["title"].value_counts() != 1]
not_oneId_title

title
Saturn 3 (1980)                           2
Confessions of a Dangerous Mind (2002)    2
Emma (1996)                               2
Eros (2004)                               2
War of the Worlds (2005)                  2
Name: count, dtype: int64

In [277]:
not_oneId_title_list = list(not_oneId_title.index)
not_oneId_title_list

['Saturn 3 (1980)',
 'Confessions of a Dangerous Mind (2002)',
 'Emma (1996)',
 'Eros (2004)',
 'War of the Worlds (2005)']

In [278]:
#แสดง row ของ title นั้น ๆ
not_oneId_movie = movie.loc[movie["title"].isin(not_oneId_title_list)].sort_values(by='title')
not_oneId_movie

movieId                                   title  \
4169    6003  Confessions of a Dangerous Mind (2002)   
9106  144606  Confessions of a Dangerous Mind (2002)   
650      838                             Emma (1996)   
5601   26958                             Emma (1996)   
5854   32600                             Eros (2004)   
9135  147002                             Eros (2004)   
2141    2851                         Saturn 3 (1980)   
9468  168358                         Saturn 3 (1980)   
5931   34048                War of the Worlds (2005)   
6932   64997                War of the Worlds (2005)   

                                   genres  
4169          Comedy|Crime|Drama|Thriller  
9106  Comedy|Crime|Drama|Romance|Thriller  
650                  Comedy|Drama|Romance  
5601                              Romance  
5854                                Drama  
9135                        Drama|Romance  
2141            Adventure|Sci-Fi|Thriller  
9468                      Sci-Fi|Thriller  
5931     Action|Adventure|Sci-Fi|Thriller  
6932                        Action|Sci-Fi

In [192]:
not_oneId_Id_list = not_oneId_movie["movieId"].unique()
not_oneId_Id_list

[6003, 144606, 838, 26958, 32600, 147002, 2851, 168358, 34048, 64997]
Categories (9742, int64): [1, 2, 3, 4, ..., 193583, 193585, 193587, 193609]

In [202]:
not_oneId_link = link.loc[link["movieId"].isin(not_oneId_Id_list)]
not_oneId_link

movieId  imdbId   tmdbId
650       838  116191   3573.0
2141     2851   81454      NaN
4169     6003  290538   4912.0
5601    26958  118308  12254.0
5854    32600  377059      NaN
5931    34048  407304     74.0
6932    64997  449040  34812.0
9106   144606  270288   4912.0
9135   147002  343663  39850.0
9468   168358   79285  19761.0

In [246]:
#เช็คว่ามีการ rating และ tag หนังที่มีปัญหามั้ย
not_oneId_rating = rating.loc[rating["movieId"].isin(not_oneId_Id_list)]
not_oneId_rating.sort_values(by='movieId')

userId  movieId  rating   timestamp
849         6      838     4.0   845555514
1332       12      838     5.0  1247264096
5096       33      838     5.0   939716327
8146       57      838     3.0   965797105
15339     100      838     4.5  1100184279
...       ...      ...     ...         ...
4747       28    64997     3.5  1234850075
11451      68    64997     2.5  1230497715
17819     111   144606     4.0  1517441257
49832     318   147002     4.0  1502207152
81458     514   168358     2.5  1533945970

[106 rows x 4 columns]

In [244]:
#เช็คว่า Id ที่หาตรงกันมั้ย
set(not_oneId_rating.movieId.unique()) == set(not_oneId_Id_list)

True

In [237]:
not_oneId_tag = tag.loc[tag["movieId"].isin(not_oneId_Id_list)]
not_oneId_tag

userId  movieId          tag   timestamp
1184     474      838  Jane Austen  1137375748
2058     474     6003   television  1138307058

### มี rating เยอะต้องจัดการ

In [205]:
left = not_oneId_link.set_index("movieId")
right = not_oneId_movie.set_index("movieId")

left.join(right).sort_values(by='title')

imdbId   tmdbId                                   title  \
movieId                                                            
6003     290538   4912.0  Confessions of a Dangerous Mind (2002)   
144606   270288   4912.0  Confessions of a Dangerous Mind (2002)   
838      116191   3573.0                             Emma (1996)   
26958    118308  12254.0                             Emma (1996)   
32600    377059      NaN                             Eros (2004)   
147002   343663  39850.0                             Eros (2004)   
2851      81454      NaN                         Saturn 3 (1980)   
168358    79285  19761.0                         Saturn 3 (1980)   
34048    407304     74.0                War of the Worlds (2005)   
64997    449040  34812.0                War of the Worlds (2005)   

                                      genres  
movieId                                       
6003             Comedy|Crime|Drama|Thriller  
144606   Comedy|Crime|Drama|Romance|Thriller  
838                     Comedy|Drama|Romance  
26958                                Romance  
32600                                  Drama  
147002                         Drama|Romance  
2851               Adventure|Sci-Fi|Thriller  
168358                       Sci-Fi|Thriller  
34048       Action|Adventure|Sci-Fi|Thriller  
64997                          Action|Sci-Fi

### เปลี่ยนให้มีแค่ Id เดียว ป้องกันการแนะนำซ้ำเรื่องเดิม

เรียงตาม title
1.Confessions of a Dangerous Mind (2002) 
-ใช้ movieId 144606 (imdb ลิ้งไปหา 270288 เสมอและ genres ครอบคลุมกว่า)
2.Emma (1996)
-ใช้ทั้งคู่เพราะเป็นหนังคนละเรื่อง คนละทีมสร้าง คนละนักแสดง และสร้างมาเพื่อคนละแพลตฟอร์ม (โรงภาพยนตร์กับโทรทัศน์)
3.Eros (2004)
-ใช้ movieId 147002 (imdb ลิ้งไปหา 343663 เสมอ, tmdb ไม่NaN และ genres ครอบคลุมกว่า)
4.Saturn 3 (1980)
-ใช้ movieId 168358 (imdb ลิ้งไปหา 79285 เสมอ)
5.War of the Worlds (2005)	
-ใช้ทั้งคู่เพราะเป็นหนังคนละเรื่อง ฉบับฟอร์มยักษ์กับเกาะกระแส

https://movielens.org/movies/"movieId"

http://www.imdb.com/title/tt0"imdbId"/

https://www.themoviedb.org/movie/"tmdbId"

### เปลี่ยน movieId ในข้อมูล user(rating กับ tag)

In [282]:
movie

movieId                                      title  \
0          1                           Toy Story (1995)   
1          2                             Jumanji (1995)   
2          3                    Grumpier Old Men (1995)   
3          4                   Waiting to Exhale (1995)   
4          5         Father of the Bride Part II (1995)   
...      ...                                        ...   
9737  193581  Black Butler: Book of the Atlantic (2017)   
9738  193583               No Game No Life: Zero (2017)   
9739  193585                               Flint (2017)   
9740  193587        Bungo Stray Dogs: Dead Apple (2018)   
9741  193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

In [301]:
# 1.Confessions of a Dangerous Mind (2002)
m1_ls = movie.loc[movie["title"] == "Confessions of a Dangerous Mind (2002)"].movieId.unique()
# 2.Emma (1996) ไม่เปลี่ยน
# 3.Eros (2004)
m3_ls = movie.loc[movie["title"] == "Eros (2004)"].movieId.unique()
# 4.Saturn 3 (1980)
m4_ls = movie.loc[movie["title"] == "Saturn 3 (1980)"].movieId.unique()
# 5.War of the Worlds (2005) ไม่เปลี่ยน

In [302]:
rating.loc[rating["movieId"].isin(m1_ls)]

userId movieId  rating   timestamp
17449    111    6003     4.0  1516468531
17819    111  144606     4.0  1517441257
23053    156    6003     3.5  1106882187
26958    182    6003     3.0  1054780821
42984    288    6003     4.0  1066059244
54020    356    6003     4.5  1229139513
59953    387    6003     3.5  1208707060
64063    414    6003     3.5  1092414917
74530    474    6003     3.5  1087831997
76779    480    6003     3.5  1179161261
85111    552    6003     3.0  1112151145
89614    580    6003     3.5  1167860103
94099    599    6003     3.0  1498527101
95721    600    6003     3.0  1237759526
99357    608    6003     4.5  1117408056
99939    610    6003     4.0  1493847220

In [303]:
rating.loc[rating["movieId"].isin(m3_ls)]

userId movieId  rating   timestamp
49832    318  147002     4.0  1502207152
98357    606   32600     3.5  1171410642

In [304]:
rating.loc[rating["movieId"].isin(m4_ls)]

userId movieId  rating   timestamp
282        3    2851     5.0  1306463925
31391    217    2851     3.0   955942393
42665    288    2851     2.0  1020369199
72810    469    2851     3.0   965335989
81458    514  168358     2.5  1533945970